File: leetCode/3.滑动窗口/3无重复字符的最长子串.js
/**
 * @param {string} s
 * @return {number}
 * abcabcbb pwwkew 3
 * bbb 1
 */
var lengthOfLongestSubstring = function(s) {
    const map = new Map()
    let max = 0
    let left = 0
    for(let i = 0; i < s.length; i++) {
        if (map.has(s[i])) {
            left = Math.max(left, map.get(s[i]) + 1)
        }
        max = Math.max(max, i - left + 1)
        map.set(s[i], i)
    }
    return max
};
console.log(lengthOfLongestSubstring('abcabcbb'))

File: leetCode/3.滑动窗口/3无重复字符的最长子串.go
func lengthOfLongestSubstring(s string) int {
    cmap := make(map[rune]int)
    left := 0
    max := 0

    for i,ch := range s {
        val, found := cmap[ch]
        if val >= 0 && found {
            left = getMax(left, cmap[ch] + 1)
        }
        max = getMax(max, i - left + 1)
        cmap[ch] = i
    }
    return max
}
func getMax(val1 int, val2 int) int {
    if (val1 >= val2) {
        return val1
    } else {
        return val2
    }
}